File: 133. Clone Graph.js
/**
 * Definition for undirected graph.
 * function UndirectedGraphNode(label) {
 *     this.label = label;
 *     this.neighbors = [];   // Array of UndirectedGraphNode
 * }
 */

/**
 * @param {UndirectedGraphNode} graph
 * @return {UndirectedGraphNode}
 */
var cloneGraph = function(graph) {
    if ( !graph ) {
        return null;
    }

    var queue = [];
    var arr = [];
    queue.push( graph );
    arr[graph.label] = new UndirectedGraphNode( graph.label );

    while ( queue.length > 0 ) {
        var node = queue.shift();
        var newNode = arr[node.label];
        if ( node.neighbors.length > 0 ) {
            for ( var i = 0; i < node.neighbors.length; i++ ) {
                if ( node.label != node.neighbors[i].label ) {
                    var neighbor = node.neighbors[i];
                    var newNeighbor;
                    if ( arr[neighbor.label] ) {
                        newNeighbor = arr[neighbor.label]
                    } else {
                        newNeighbor = new UndirectedGraphNode( node.neighbors[i].label );
                        queue.push( node.neighbors[i] );
                        arr[node.neighbors[i].label] = newNeighbor;
                    }
                    newNode.neighbors.push( newNeighbor );
                } else {
                    newNode.neighbors.push( arr[node.label] );
                }
            }
        }
    }
    return arr[graph.label];
};

File: 133. Clone Graph.py
# Definition for a undirected graph node
# class UndirectedGraphNode:
#     def __init__(self, x):
#         self.label = x
#         self.neighbors = []

"""
将graph上的每个node都放到map上
"""
class Solution:
    # @param node, a undirected graph node
    # @return a undirected graph node
    def cloneGraph(self, node):
        if node is None:
            return node

        root = node
        import Queue
        q = collections.deque([node])
        nodes = set([node])
        while q:
            head = q.popleft()
            for neighbor in head.neighbors:
                if neighbor not in nodes:
                    nodes.add(neighbor)
                    q.append(neighbor)

        mapping = {}
        for node in nodes:
            mapping[node] = UndirectedGraphNode(node.label)

        for node in nodes:
            newNode = mapping[node]
            for neighbor in node.neighbors:
                newNeighbor = mapping[neighbor]
                newNode.neighbors.append(newNeighbor)

        return mapping[root]
